File: src/reasoning/confidence.ts
import type { ConfidenceConfig, IterationQualityMetrics, RDTIterationState } from '../types/rdt.types.js';

function clamp(value: number, min = 0, max = 1): number {
  return Math.max(min, Math.min(max, value));
}

function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, ' ')
    .split(/\s+/)
    .filter((token) => token.length > 2);
}

function overlapScore(a: string[], b: string[]): number {
  if (a.length === 0 || b.length === 0) return 0.35;
  const aSet = new Set(a);
  const bSet = new Set(b);
  const intersection = [...aSet].filter((token) => bSet.has(token)).length;
  const union = new Set([...aSet, ...bSet]).size;
  return union === 0 ? 0 : intersection / union;
}

function containsConclusion(summary: string): boolean {
  const markers = ['therefore', 'final', 'answer', 'conclusion', 'thus', 'result'];
  const lower = summary.toLowerCase();
  return markers.some((marker) => lower.includes(marker));
}

function contradictionHint(summary: string): number {
  const lower = summary.toLowerCase();
  const weakSignals = ['maybe', 'unsure', 'not certain', 'unclear'];
  const conflictSignals = ['however', 'but', 'on the other hand', 'contradict'];
  const weakPenalty = weakSignals.some((signal) => lower.includes(signal)) ? 0.12 : 0;
  const conflictPenalty = conflictSignals.some((signal) => lower.includes(signal)) ? 0.08 : 0;
  return clamp(1 - (weakPenalty + conflictPenalty));
}

export interface ConfidenceAssessment {
  confidence: number;
  threshold: number;
  shouldExit: boolean;
  shouldRevise: boolean;
  metrics: IterationQualityMetrics;
}

export class ConfidenceEvaluator {
  private adaptiveThreshold: number;
  private readonly history: Array<{ confidence: number; threshold: number; metrics: IterationQualityMetrics }> = [];

  constructor(private readonly config: ConfidenceConfig) {
    this.adaptiveThreshold = clamp(config.thresholds.earlyExit);
  }

  evaluate(current: string, prior: RDTIterationState | undefined, decomposition: string[]): ConfidenceAssessment {
    const priorSummary = prior?.summary ?? '';
    const currentTokens = tokenize(current);
    const priorTokens = tokenize(priorSummary);

    const coherence = clamp(0.45 + overlapScore(currentTokens, priorTokens) * 0.55);
    const decompositionCoverage = decomposition.length === 0
      ? 1
      : decomposition.filter((step) => current.toLowerCase().includes(step.toLowerCase().slice(0, 24))).length / decomposition.length;
    const sizeAdequacy = clamp(Math.min(1, current.length / 220));
    const completeness = clamp(sizeAdequacy * 0.5 + decompositionCoverage * 0.35 + (containsConclusion(current) ? 0.15 : 0));
    const consistency = contradictionHint(current);

    const aggregate = clamp(coherence * 0.4 + completeness * 0.35 + consistency * 0.25);
    const confidence = this.smoothConfidence(aggregate, prior?.confidence);

    if (this.config.adaptive) {
      this.adaptThreshold(confidence, prior?.confidence ?? confidence);
    }

    const threshold = this.adaptiveThreshold;
    const metrics: IterationQualityMetrics = {
      coherence,
      completeness,
      consistency,
      aggregate,
      explanation: `coherence=${coherence.toFixed(2)}, completeness=${completeness.toFixed(2)}, consistency=${consistency.toFixed(2)}`
    };

    const shouldExit = confidence >= threshold;
    const shouldRevise = confidence < this.config.thresholds.revise;

    this.history.push({ confidence, threshold, metrics });

    return {
      confidence,
      threshold,
      shouldExit,
      shouldRevise,
      metrics
    };
  }

  getHistory(): Array<{ confidence: number; threshold: number; metrics: IterationQualityMetrics }> {
    return [...this.history];
  }

  getAdaptiveThreshold(): number {
    return this.adaptiveThreshold;
  }

  private smoothConfidence(raw: number, prior?: number): number {
    if (prior === undefined) return clamp(raw);
    return clamp(prior * this.config.smoothingFactor + raw * (1 - this.config.smoothingFactor));
  }

  private adaptThreshold(current: number, previous: number): void {
    if (current > previous + 0.08) {
      this.adaptiveThreshold = clamp(this.adaptiveThreshold + this.config.adaptUpDelta, this.config.thresholds.floor, 0.98);
      return;
    }

    if (current < previous - 0.08) {
      this.adaptiveThreshold = clamp(this.adaptiveThreshold - this.config.adaptDownDelta, this.config.thresholds.floor, 0.98);
    }
  }
}

File: backend/zygos/reasoning/confidence.py
"""Local, zero-LLM confidence heuristic (ported + cleaned from v1).

coherence  - token-overlap stability vs the prior summary
completeness - size adequacy + decomposition coverage + conclusion marker
consistency  - contradiction/uncertainty penalty
aggregate  - weighted blend, smoothed across iterations
Stability: Experimental.
"""

import re

from zygos.reasoning.profiles import ResolvedProfile
from zygos.reasoning.types import ConfidenceBreakdown

_CONCLUSION = ("therefore", "final", "answer", "conclusion", "thus", "result")
_WEAK = ("maybe", "unsure", "not certain", "unclear")
_CONFLICT = ("however", "but", "on the other hand", "contradict")
_NON_ALNUM = re.compile(r"[^a-z0-9]+")


def _clamp(x: float) -> float:
    return max(0.0, min(1.0, x))


def _tokens(text: str) -> set[str]:
    return {t for t in _NON_ALNUM.sub(" ", text.lower()).split() if len(t) > 2}


def _jaccard(a: set[str], b: set[str]) -> float:
    if not a or not b:
        return 0.35
    return len(a & b) / len(a | b)


def _coherence(current: str, prior: str | None) -> float:
    return _clamp(0.45 + _jaccard(_tokens(current), _tokens(prior or "")) * 0.55)


def _decomposition_coverage(current: str, decomposition: tuple[str, ...]) -> float:
    if not decomposition:
        return 1.0
    low = current.lower()
    hits = sum(1 for step in decomposition if step[:24].lower() in low)
    return hits / len(decomposition)


def _completeness(current: str, decomposition: tuple[str, ...]) -> float:
    size = _clamp(min(1.0, len(current) / 220))
    coverage = _decomposition_coverage(current, decomposition)
    low = current.lower()
    conclusion = 0.15 if any(marker in low for marker in _CONCLUSION) else 0.0
    return _clamp(size * 0.5 + coverage * 0.35 + conclusion)


def _consistency(current: str) -> float:
    low = current.lower()
    weak = 0.12 if any(marker in low for marker in _WEAK) else 0.0
    conflict = 0.08 if any(marker in low for marker in _CONFLICT) else 0.0
    return _clamp(1.0 - (weak + conflict))


def _adapt_threshold(
    prev_threshold: float, current_raw: float, prior_aggregate: float | None, profile: ResolvedProfile
) -> float:
    if prior_aggregate is None:
        return profile.early_exit
    if current_raw > prior_aggregate + 0.08:
        return min(0.98, prev_threshold + profile.adapt_up)
    if current_raw < prior_aggregate - 0.08:
        return max(profile.floor, prev_threshold - profile.adapt_down)
    return prev_threshold


def score(
    current: str,
    prior_summary: str | None,
    decomposition: tuple[str, ...],
    prior_aggregate: float | None,
    prev_threshold: float,
    profile: ResolvedProfile,
) -> ConfidenceBreakdown:
    coherence = _coherence(current, prior_summary)
    completeness = _completeness(current, decomposition)
    consistency = _consistency(current)
    raw = _clamp(coherence * 0.4 + completeness * 0.35 + consistency * 0.25)
    if prior_aggregate is None:
        aggregate = raw
    else:
        aggregate = _clamp(prior_aggregate * profile.smoothing + raw * (1 - profile.smoothing))
    threshold = _adapt_threshold(prev_threshold, raw, prior_aggregate, profile)
    return ConfidenceBreakdown(
        coherence=coherence, completeness=completeness, consistency=consistency,
        raw_aggregate=raw, aggregate=aggregate, threshold=threshold,
    )


def in_fence_band(aggregate: float, threshold: float, band: float) -> bool:
    return abs(aggregate - threshold) <= band
